cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/spark/SparkProgramRunnerTest.java

```diff
@@ -396,6 +400,122 @@ private void testSparkWithCustomFileSet(Class<?> appClass, Class<?> programClass
+  }
+
+  private void testSparkWithGetDataset(Class<?> appClass, Class<?> programClass) throws Exception {
+    ApplicationWithPrograms app =
+      AppFabricTestHelper.deployApplicationWithManager(appClass, TEMP_FOLDER_SUPPLIER);
+    FileSet fileset = datasetInstantiator.getDataset("logs");
+    Location location = fileset.getLocation("nn");
+    prepareInputFileSetWithLogData(location);
+
+    Map<String, String> inputArgs = new HashMap<>();
+    FileSetArguments.setInputPath(inputArgs, "nn");
+    Map<String, String> args = new HashMap<>();
+    args.putAll(RuntimeArguments.addScope(Scope.DATASET, "logs", inputArgs));
+    args.put("input", "logs");
+    args.put("output", "logStats");
+
+    runProgram(app, programClass, args);
+
+    KeyValueTable logStatsTable = datasetInstantiator.getDataset("logStats");
+    validateGetDatasetOutput(logStatsTable);
+  }
+
+  private void validateGetDatasetOutput(KeyValueTable logStatsTable) {
+    CloseableIterator<KeyValue<byte[], byte[]>> scan = logStatsTable.scan(null, null);
+    try {
+      // must have 4 records
+      Assert.assertTrue(scan.hasNext());
+      validateKeyValue(scan.next());
+      Assert.assertTrue(scan.hasNext());
+      validateKeyValue(scan.next());
+      Assert.assertTrue(scan.hasNext());
+      validateKeyValue(scan.next());
+      Assert.assertTrue(scan.hasNext());
+      validateKeyValue(scan.next());
+      // no more records
+      Assert.assertFalse(scan.hasNext());
+    } finally {
+      scan.close();
+    }
+  }
+
+  private void validateKeyValue(KeyValue<byte[], byte[]> keyValue) {
+    SparkAppUsingGetDataset.LogKey fredKey1 = new SparkAppUsingGetDataset.LogKey(
+        "10.10.10.10", "FRED", "GET http://bar.com/image.jpg HTTP/1.1", 200);
+    SparkAppUsingGetDataset.LogKey fredKey2 = new SparkAppUsingGetDataset.LogKey(
+        "10.10.10.10", "FRED", "GET http://bar.com/image.jpg HTTP/1.1", 404);
+    SparkAppUsingGetDataset.LogKey bradKey1 = new SparkAppUsingGetDataset.LogKey(
+        "20.20.20.20", "BRAD", "GET http://bar.com/image.jpg HTTP/1.1", 200);
+    SparkAppUsingGetDataset.LogKey bradKey2 = new SparkAppUsingGetDataset.LogKey(
+        "20.20.20.20", "BRAD", "GET http://bar.com/image.jpg HTTP/1.1", 404);
+    SparkAppUsingGetDataset.LogStats fredStats1 = new SparkAppUsingGetDataset.LogStats(2, 100);
+    SparkAppUsingGetDataset.LogStats fredStats2 = new SparkAppUsingGetDataset.LogStats(1, 50);
+    SparkAppUsingGetDataset.LogStats bradStats1 = new SparkAppUsingGetDataset.LogStats(1, 50);
+    SparkAppUsingGetDataset.LogStats bradStats2 = new SparkAppUsingGetDataset.LogStats(1, 50);
+
+    SparkAppUsingGetDataset.LogKey logKey =
+      GSON.fromJson(Bytes.toString(keyValue.getKey()), SparkAppUsingGetDataset.LogKey.class);
+    SparkAppUsingGetDataset.LogStats logStats =
+      GSON.fromJson(Bytes.toString(keyValue.getValue()), SparkAppUsingGetDataset.LogStats.class);
+
+    if (fredKey1.equals(logKey)) {
```

Comment:
I think this if else loop can be simplified a bit if you put the above in a Map and then do contains key with logKey in then get if found to compare value else your assert fail.